include/dlaf/sender/transform_mpi.h

```diff
@@ -70,17 +74,41 @@ struct MPICallHelper {
     if constexpr (std::is_void_v<result_type>) {
       std::move(f)(dlaf::common::internal::unwrap(ts)..., &req);
       (internal::consumeCommunicatorWrapper(ts), ...);
-      pika::util::yield_while(is_request_completed);
+      pika::util::yield_while([req]() { return !mpid::poll_request(req); });
     }
     else {
-      auto r = std::move(f)(dlaf::common::internal::unwrap(ts)..., &req);
+      /*auto r = */ std::move(f)(dlaf::common::internal::unwrap(ts)..., &req);
+      (internal::consumeCommunicatorWrapper(ts), ...);
+      pika::util::yield_while([req]() { return !mpid::poll_request(req); });
+    }
+  }
+};
+
+/// Helper type for wrapping MPI calls.
+template <typename F>
+struct MPICallHelper {
+  std::decay_t<F> f;
+
+  template <typename... Ts>
+  auto operator()(Ts&&... ts) -> decltype(std::move(f)(dlaf::common::internal::unwrap(ts)...)) {
+    using namespace pika::debug::detail;
+    PIKA_DETAIL_DP(dla_debug<5>, debug(str<>("MPICallHelper"), pika::debug::print_type<Ts...>(", ")));
+    using result_type = decltype(std::move(f)(dlaf::common::internal::unwrap(ts)...));
+    if constexpr (std::is_void_v<result_type>) {
+      std::move(f)(dlaf::common::internal::unwrap(ts)...);
+      (internal::consumeCommunicatorWrapper(ts), ...);
+    }
+    else {
+      auto r = std::move(f)(dlaf::common::internal::unwrap(ts)...);
       (internal::consumeCommunicatorWrapper(ts), ...);
-      pika::util::yield_while(is_request_completed);
       return r;
```

Comment:
If we had a scope guard type of thing (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/n4853.html#scopeguard.exit) in DLA-Future we could write this roughly as:
```
      scope_exit s{[&] { (internal::consumeCommunicatorWrapper(ts), ...); });  
      return std::move(f)(dlaf::common::internal::unwrap(ts)...);
```
without the need to branch for void and non-void cases.

I'm not expecting a change here, but I just wanted to write this down here as a reminder for when we get closer to merging this PR.